### ISSUE TITLE
Add support for user categories

### ISSUE THREAD

**Issue by cosmicchipsocket** [tkashkin/GameHub#47]
It would be nice to be able to create categories and use them to organize your games. There are several use cases for this:

- Making a list of favorite games
- Tracking which games you've already completed
- Hiding games that don't run on your machine (example: Anomaly: Warzone Earth is known to be broken on Linux yet it appears near the top of the list due to alphabetical sorting)
- Hiding games you got as part of a bundle but aren't interested in playing

Being able to show/hide these categories would also help cut down on clutter, and you should be able to add a game to multiple categories (as if they were tags).

The native Steam client has this functionality, and I think it's worth having here too.

**Comment by tkashkin** [comment]
@cosmicchipsocket what do you think about something like this?

![filters_mockup](https://user-images.githubusercontent.com/2003017/44626829-4bca3b80-a92c-11e8-9cf7-92f593abfa64.png)

I also need some icon for filters popup. Some sort of tag icon may be appropriate, but what if this popup will later have other filters?

**Comment by tkashkin** [comment]
This layout was the first thing that I've thought of, but I agree that it may not be a good solution.
Probably checklist will be better.

Do you suggest Steam-like headers for tags. I don't really like this solution, because expected behaviour in this case is to list game in each tag it has, and it may become too much.

Example: I have used Depressurizer to import Steam tags as library categories. I have many categories. This is a VR section with just 11 games and it's already a mess if games have multiple tags.

![image](https://user-images.githubusercontent.com/2003017/44631928-93c47f00-a97b-11e8-9230-97e1a77d76fc.png)

**Comment by tkashkin** [comment]
@neuromancer, @MessedUpHare, do you have any ideas or suggestions?

**Comment by neuromancer** [comment]
I think you can add [IGDB](https://www.igdb.com) support and extract keywords from there as well.

**Comment by cosmicchipsocket** [comment]
If there's automatic tagging, I'd prefer that it be optional, and placed separately in the list from user-made categories.

Maybe have a settings dialog for managing the categories, where you can rename/delete user categories, and choose between providers for automatic ones?

**Comment by tkashkin** [comment]
What do you think about this?

![image](https://user-images.githubusercontent.com/2003017/44638577-48808f80-a9c0-11e8-9c88-856bbee61f3a.png)

The first two tags (favorites and hidden) are builtin non-removable tags.
The rest of them are imported from GOG. They are created by user but GOG can probably have some predefined ones.

**Comment by tkashkin** [comment]
@cosmicchipsocket, @neuromancer, @MessedUpHare, tags are working now (no user-created tags yet).

GOG tags will be imported. Games can be marked as favorite or hidden with right click in both grid and list views.

**Comment by neuromancer** [comment]
I'm afraid that last revision is broken for me. It does not load any game, I believe is related with #48, but it does not seem to be fixed changing some options (and unfortunately, I won't be able to submit a proper issue report in two weeks).

**Comment by tkashkin** [comment]
@neuromancer I'm still not sure what's the problem as it fixed itself on my laptop. Have you waited long enough to get all your games cached in db?

**Comment by neuromancer** [comment]
GOG categories are working, but the icon of that menu is missing:

![missing-icon](https://user-images.githubusercontent.com/110782/45209986-b2394d00-b28e-11e8-8ff9-49160c53c879.png)

(I'm using 53e10bbc7f869151cc9ff0fd08d44338774ad80d )

**Comment by tkashkin** [comment]
@cosmicchipsocket check with the latest commit in dev branch. Do you have any ideas or suggestions?

**Comment by tkashkin** [comment]
> Went ahead and installed the latest dev build.

Right now `master` branch is more recent than `dev` (but most of the difference are fixes for AppImage and building on 16.04).

> There's also the issue of every game being listed as "Run with compatibility layer" but that's an unrelated issue.

Yes, I know about that bug.

> I'd expect custom tags to have their own headers

What header should be displayed if game have multiple tags?

> I'm not sure why I can select GOG tags for games on Steam.

I'm not sure why it should be limited. GOG tags are the same as any other tags, all the difference is that GOG tags are auto-imported from GOG and have GOG icon.

> Bug report: Making a new tag named "Complete" and then hiding it will instead hide the "Not installed" category.

I see that, not sure yet why that happens. It doesn't hide all of the not installed games, but some of them.
There's now quite [complicated logic in filtering](https://github.com/tkashkin/GameHub/blob/0.11.1-2-master/src/ui/views/GamesView/GamesView.vala#L513-L571), there's a very high chance it won't handle some cases correctly.
One possible way to fix this is to refactor filtering to use only tags and implement all of the current conditions (except search) as builtin tags (like favorites and hidden currently are).

**Comment by tkashkin** [comment]
@cosmicchipsocket 

> Bug report: Making a new tag named "Complete" and then hiding it will instead hide the "Not installed" category.

Should be fixed in `dev` branch ([0.11.2-1-dev](https://github.com/tkashkin/GameHub/releases/tag/0.11.2-1-dev)). It actually worked correctly, but broke after adding  "Installed" tag (#69). Now it should work properly with "Not installed" tag added.

**Comment by cosmicchipsocket** [comment]
> I'm not sure why it should be limited. GOG tags are the same as any other tags, all the difference is that GOG tags are auto-imported from GOG and have GOG icon.

Because the GOG icon on them implies that they are synchronized with GOG and it'd be hard to conceptualize GOG knowing how to apply a tag to a Steam game.

(At the very least I'd like to be able to disable the GOG tags)

**Comment by tkashkin** [comment]
@cosmicchipsocket imported tags can be disabled since cb4c384.

**Comment by Jimi-James** [comment]
Are there any plans to support importing Steam's categories as tags? I have a major amount of categories set up, too, and I'd like to have them totally synced between GameHub and Steam. Doing that manually would be too ridiculous.

@tkashkin You mentioned importing Steam's categories in with Depressurizer earlier? How exactly did you do that? I'm only just now hearing about Depressurizer, and it seems to be Windows-only. I'm on Linux. I do have a Windows VM, but if I used it, is there actually a way through Depressurizer to, I guess, export Steam's categories as a format that's easy to plug into GameHub's config files, and then copy those config files into Linux? I mean, that's way too roundabout to be a permanent workaround. However, I wouldn't mind doing it this first time until more of a sync situation can be implemented.

**Comment by tkashkin** [comment]
@Jimi-James 

> Are there any plans to support importing Steam's categories as tags?

No plans right now, but it should be doable. I'll need to research how does Steam store user-added categories.

> You mentioned importing Steam's categories in with Depressurizer earlier?

I have mentioned it only as an example of what a mess it would be if there was a list header for each tag and game had multiple tags.

**Comment by Jimi-James** [comment]
@tkashkin 
> No plans right now, but it should be doable. I'll need to research how does Steam store user-added categories.

Oh, I'm pretty sure that's in `<steam>/userdata/<UID>/7/remote/sharedconfig.vdf`. Actually, I just checked my sharedconfig.vdf and, yeah, all my custom categories organized by app ID, and I think also all my custom shortcuts and their categories, were in that file. Ooh, importing custom shortcuts would be cool, too, though probably much more complicated.

> I have mentioned it only as an example of what a mess it would be if there was a list header for each tag and game had multiple tags.

Oh, wait, I just realized you actually imported Steam *community* tags. As in not user categories. OK, yeah, I see, that wasn't what I was hoping at all. Oh well.

**Comment by Jimi-James** [comment]
After all the bugfixing we just did elsewhere, GameHub is almost flawless for my setup! All that's left to do now is get my Steam categories and non-Steam game shortcuts in a sync setup with GameHub.

Which I've figured out the best way to accomplish is on my own, in a launcher-agnostic way (I've lived through Ice and Steam-ROM-Maneger each working and then stopping working forever, forgetting their settings now and then along the way, and I'm sick of that). It looks like I'll be able to build a launcher-agnostic syncing setup using only Bash scripts that I figure out how to plug into GameHub's config files, and a way to parse Steam's infamous `shortcuts.vdf`.

Turns out there's not only a Python library for that (https://pypi.org/project/vdf/), but also a complete project someone else has already "finished" (it's a bit out-of-date now: https://github.com/CorporalQuesadilla/Steam-Shortcut-Manager), who I'm now talking to over reddit. When I have more time later, I'll probably help with that, and either way, once it's all back in working order, I can put what I figure out about GameHub integration into its wiki, and then post back here with that whole setup!

**Comment by tkashkin** [comment]
@Jimi-James GameHub can already read, change and save `shortcuts.vdf`. It supports adding games to Steam as non-Steam games. https://github.com/tkashkin/GameHub/issues/149#issuecomment-445139556

Now it adds games to `shortcuts.vdf` with one tag ("GameHub").

It should be possible to extend it to allow two-way sync.

**Comment by Jimi-James** [comment]
Oh! I... OK, I just failed to find that setting myself. I guess I didn't realize it was already there because I didn't try adding a shortcut *through* GameHub yet (as I had actually just finished dramatically reorganizing my Steam library immediately before finding GameHub)? And because the two-way sync hasn't been implemented yet? So when that happens, everything will just show up in here? To be honest, I'd rather wait for two-way sync to be implemented, and then when I have time and if it hasn't happened yet, try helping with a pull request, than set up the same custom shortcuts and categories all over again in a new program.

**Comment by tkashkin** [comment]
> Does the current implementation already use, or have an option to use, steam://rungameid/ to run those custom shortcuts through Steam, so that you can get an in-game status, the overlay, and controller configs and such?

No, it does not. This is an interesting idea, but I'm not sure yet how to implement it.

Current supported scenario is a bit different:
* Install GOG/Humble/etc game through GameHub
* Add shortcut to Steam via "Add to Steam library" button in game properties
* Use Steam Big Picture or desktop client as a launcher and it will launch game through GameHub

**Comment by Jimi-James** [comment]
Alright, that's what I figured, as it doesn't seem to work with native games yet. If I try to make just a shortcut to a binary on my system, I get a shortcut that doesn't actually do anything. It lists itself as Not installed, and trying to install it does nothing. Clicking Add to steam does nothing. On top of that, it won't even remember the binary that it's set to use.

**Comment by tkashkin** [comment]
@Jimi-James that's weird. It works fine for me:

![screenshot](https://user-images.githubusercontent.com/2003017/55281084-5c1e1b00-5340-11e9-9838-2e06f26ef049.png)

Added shortcut:
```
  |- [3]
    |- StartDir: .
    |- LaunchOptions: --run user:b5eccde58b6cca7892b38a4c0a7a7ec5
    |- OpenVR: 0
    |- AllowOverlay: 1
    |- AllowDesktopConfig: 1
    |- [tags]
      |- 0: GameHub
    |- exe: com.github.tkashkin.gamehub
    |- LastPlayTime: 1
    |- AppName: QZDoom
    |- IsHidden: 0
    |- ShortcutPath: /usr/share/applications/com.github.tkashkin.gamehub.desktop
```

Run GameHub from terminal with at least `--debug` argument and try to do that again.

**Comment by Jimi-James** [comment]
The plot thickened--I took a while to notice (I've been in a busy spell), but Steam actually did get my Test shortcut added to it. Several times--once for each time I pressed the button, only to be confused by the lack of any feedback. So, the only thing that's actually not working there is I can't set a binary or a starting path--both always become blank again as soon as I close and reopen that shortcut's config window. Weirdly, a binary (Target) and starting path do end up being set in the corresponding Steam shortcut, but they're not exactly what I had in mind:

![Screenshot_2019-04-02_10-00-32](https://user-images.githubusercontent.com/5751606/55421577-600f9000-552e-11e9-9cd3-6f19d2affe59.png)

When I'm less busy again (probably later this week), I'll try all this again from scratch with the `--debug` option.
